File: homework3/phuc_le/q8_AlternatingPath.cpp
#include <iostream>
#include <vector>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <queue>
#include <tuple>

using namespace std;

// Create an Edge struct to represent the graph edges
struct Edge
{
    string src;
    string dest;
    string color;
};

// Custom hash function so we can store std::pair<string, string> in an unordered_set
struct pair_hash
{
    inline size_t operator()(const pair<string, string> &p) const
    {
        return hash<string>()(p.first) ^ hash<string>()(p.second);
    }
};

int alternatingPath(const vector<Edge> &edges, const string &start, const string &end)
{
    if (start == end)
    {
        return 0; // No edges needed if start and end are the same
    }

    // Queue stores tuples of: (Current Node, Color of the edge used to get here, Current Distance)
    queue<tuple<string, string, int>> q;
    // Build the graph as an adjacency list: node -> list of (neighbor, color)
    unordered_map<string, vector<pair<string, string>>> graph;
    for (const auto &edge : edges)
    {
        graph[edge.src].push_back({edge.dest, edge.color});
    }

    // Visited set to track (node, lastColor) pairs to prevent cycles
    unordered_set<pair<string, string>, pair_hash> visited;

    // Starting color can be any color
    q.push({start, "", 0}); // (currentNode, lastColor, pathLength)
    visited.insert({start, ""});

    while (!q.empty())
    {
        auto [currentNode, lastColor, pathLength] = q.front();
        q.pop();

        if (currentNode == end)
        {
            return pathLength; // Found the end node, return the path length
        }

        // Check all neighbors
        for (const auto &neighbor : graph[currentNode])
        {
            string nextNode = neighbor.first;
            string edgeColor = neighbor.second;

            if (lastColor.empty() || edgeColor != lastColor)
            {
                pair<string, string> state = {nextNode, edgeColor};
                if (visited.find(state) == visited.end())
                {
                    visited.insert(state);
                    q.push({nextNode, edgeColor, pathLength + 1});
                }
            }
        }
    }
    return -1; // No alternating path found
}

// Helper function to run tests and format output cleanly
void runTest(int testNum, const vector<Edge> &edges, const string &origin, const string &dest, int expected)
{
    int result = alternatingPath(edges, origin, dest);
    cout << "Test " << testNum << " (Origin: " << origin << ", Dest: " << dest << "): ";
    if (result == expected)
    {
        cout << "PASS (Distance: " << result << ")\n";
    }
    else
    {
        cout << "FAIL\n";
    }
    cout << "  Expected: " << expected << "\n";
    cout << "  Actual:   " << result << "\n";
    cout << "\n";
}

int main()
{
    cout << "--- Running AlternatingPath Test Suite ---\n\n";

    vector<Edge> graphEdges = {
        {"A", "B", "blue"},
        {"A", "C", "red"},
        {"B", "D", "blue"},
        {"B", "E", "blue"},
        {"C", "B", "red"},
        {"D", "C", "blue"},
        {"A", "D", "red"},
        {"D", "E", "red"},
        {"E", "C", "red"}};

    // 1. Example 1
    runTest(1, graphEdges, "A", "E", 4);

    // 2. Example 2 (Unreachable via alternating colors)
    runTest(2, graphEdges, "E", "D", -1);

    // 3. Edge Case: Origin is Destination
    runTest(3, graphEdges, "A", "A", 0);

    // 4. Edge Case: Reachable but no alternating path (C -> E)
    // C to B is red. B to E is blue. Total distance 2.
    runTest(4, graphEdges, "C", "E", 2);

    return 0;
}

File: homework3/aman_varshney/q8_AlternatingPath.py
# spent 45 minutes
# bfs 
# TC - O(V+E)
# SC - O(V) for visited set and O(V+E) for adjacency list

from collections import deque


def alternatingPaths(source: str, destination: str, graph: list[tuple[str, str, str]]) -> int:
    # build adjacency list
    adj = {} # source : (destination, color)
    for src, dest, color in graph:
        if src in adj:  
            adj[src].append((dest, color))
        else:
            adj[src] = [(dest, color)]
            
        
    # bfs queue
    queue = deque() # (node, last color, distance)
    queue.append( (source, None, 0) )
    # visited set  
    visited = set() # (node, last color)
    visited.add( (source, None) )
    
    while queue:
        node, last_color, dist = queue.popleft()
        
        if node == destination: # found
            return dist
        
        for nei, color in adj[node]: # check adjacency list
            if color != last_color: # enqueue only if alternating path
                state = (nei, color) 
                if state not in visited: # enqueue only if new path
                    visited.add(state)
                    queue.append( (nei, color, dist+1) )
    
    return -1 # no alternating path


    
    
    
if __name__ == "__main__":
    graph = [("A", "B", "blue"), ("A", "C", "red"), ("B", "D", "blue"), ("B", "E", "blue"), ("C", "B", "red"), ("D", "C", "blue"), ("A", "D", "red"), ("D", "E", "red"), ("E", "C", "red")]
    src1 = "A"
    dest1 = "E"
    print("Expected: 4")
    print("Actual:", alternating_path(src1, dest1, graph))
    
    src2 = "E"
    dest2 = "D"
    print("Expected: -1")
    print("Actual", alternating_path(src2, dest2, graph))
        
    
